### ISSUE TITLE
About general activation function

### ISSUE THREAD

**Issue by luzai** [mapillary/inplace_abn#77]
Great thanks for inplace_abn, it increase batch size by 1.5 in my application! 

I notice that `InPlaceABN` support  `no activatation function`, `relu`, `elu`, `leaky relu` now. May I ask how to implement other activation function like prelu? 

Can we just use `[InPlaceABN(depth, activation='none'), torch.nn.PReLU()]`  , since PReLU seems not inplace operation?

**Comment by rotabulo** [comment]
@luzai you can render PReLU inplace as well, as long as you prevent the network to put the learned slope to 0. In principle, one can implement PReLU by modifying our code and compute the derivative with respect to the slope, which can be computed given the inverted PReLU.

**Comment by luzai** [comment]
Thank you very much for you advice! I try to render PReLU inplace later. 

I first adopt `[InPlaceABN(depth, activation='none'), torch.nn.PReLU()]`. Since I have trained a model with  `[torch.nn.BatchNorm2d(depth), torch.nn.PReLU()]`, I want to load the state_dict into new model. But I find the validation accuracy drops from 94% to 88%. 

May I ask how to load the pretrained model? These two batch norm seems not the same.  
``` python 
inp = to_torch(np.random.rand(1,64,112,112)) .float()
weight = to_torch(np.random.rand( 64 )).float()
bias = to_torch(np.random.rand( 64 )).float()
rmean = to_torch(np.random.rand( 64 )).float()
rvar = to_torch(np.random.rand( 64 )).float() 
bn1 = InPlaceABN(64,activation='none')
bn2 = torch.nn.BatchNorm2d(64)
bn1.weight.data = weight.clone() 
bn1.bias.data = bias.clone() 
bn1.running_mean=rmean.clone() 
bn1.running_var = rvar.clone() 

bn2.weight.data = weight.clone() 
bn2.bias.data = bias.clone() 
bn2.running_mean=rmean.clone() 
bn2.running_var = rvar.clone() 

bn1.eval()
bn2.eval()

((bn1(inp)-bn2(inp))**2).sum()
```

**Comment by rotabulo** [comment]
@luzai the difference in running time between Pytorch's BN and our implementation is due to the fact that our code is not as optimized as the cudnn implementation of BN that Pytorch uses.

The difference in the output is due to a change that we introduced after publishing our paper, which changes the way the scale parameter of BN is used. In our implementation the scale `a` (parameter `weight`  in Pytorch's BN) is replaced with `a' = eps + |a|`$, where `eps` is a small positive constant and `| |` is the absolute value function. The reason for this change was to prevent $\alpha$ to become zero, ensuring invertiblity. This change might performance when initialized from models pretrained with standard BN in case the scale parameter of the latter models is negative. Is this the case with your models?

**Comment by luzai** [comment]
Great thanks for your comments! 

I understand that inpabn ([with the change of scale `a`](https://github.com/pytorch/pytorch/blob/866c4e3467de241fe2c99f0f39f33d3f60c7d4c2/torch/csrc/api/src/nn/modules/batchnorm.cpp#L64)) initialized from standard BN might meets performance degradation, but standard BN initialized from ipabn just meets some randomization between cudnn and cuda, since `a` of ipabn is positive.

**Comment by rotabulo** [comment]
@luzai yes, using iabn initialized from weights trained with standard BN might lead to performance degradation in case you have negative scaling factors. Instead, I would expect minor differences if you apply standard bn to iabn weights. For the former case, one workaround could be to change the sign of some weights of the preceding conv layer in order to invert the sign of the channels having the negative scaling factors, which in turn can be taken with positive sign. If you go this way and you want also to pre-initialize the running statistics, you need to change sign also to the running means.

**Comment by luzai** [comment]
Thank you very much for your advice! As mentioned in your paper, "at test-time BN becomes a linear operation". It is quite inspiring and account for the reason that changing both sign leads to same results. I will try to change the sign of the weight/bias of conv, and of the running means.

**Comment by twmht** [comment]
@luzai 

```
I first adopt [InPlaceABN(depth, activation='none'), torch.nn.PReLU()]. Since I have trained a model with [torch.nn.BatchNorm2d(depth), torch.nn.PReLU()], I want to load the state_dict into new model. But I find the validation accuracy drops from 94% to 88%.
```

Did you solve the accuracy drop problem?

**Comment by twmht** [comment]
@rotabulo 

Since the forward implementation is different from the standard one, so we have to fuse convolution and batch normalization before we convert the model weight to onnx for deployment?

**Comment by rotabulo** [comment]
@twmht it should be possible to remap `InplaceABN` to `ABN` (i.e. BN + Activation) by simply remapping `weight`  to `weight.abs_() + eps`, where `esp=1e-5` unless you changed the default value

**Comment by poornimajd** [comment]
@twmht @rotabulo While converting my torch model which uses inplace abn, to onnx I face the issue that inplace abn is not supported in onnx,is there any workaround for this?
The model has been trained with inplaceabn and now I want to exported the trained model to onnx.
Any suggestion is greatly appreciated.
Thank you

**Comment by twmht** [comment]
@poornimajd 

you need to convert inplace abn to bn before converting to onnx.

as suggested

> it should be possible to remap InplaceABN to ABN (i.e. BN + Activation) by simply remapping weight to weight.abs_() + eps, where esp=1e-5 unless you changed the default value

**Comment by poornimajd** [comment]
@twmht  Thank you for the suggestion.But if it is possible can you just tell me how do I change this weight to weight.abs_() + eps, in the code,or is it while loading the weights?
